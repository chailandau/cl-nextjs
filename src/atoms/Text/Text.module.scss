.text {
    @include transition(font-size);
    &.xs {
        @include font(16, 28);
        &.rich-text + &.rich-text {
            margin-top: rem(24);
        }
    }
    &.sm {
        @include font(16, 28);
        &.rich-text + &.rich-text {
            margin-top: rem(28);
        }
        @include tablet {
            @include font(20, 30);
        }
    }
    &.md {
        @include font(20, 30);
        &.rich-text + &.rich-text {
            margin-top: rem(30);
        }
        @include tablet {
            @include font(22, 32);
        }
    }
    &.lg {
        @include font(24, 48);
        &.rich-text + &.rich-text {
            margin-top: rem(32);
        }
        @include tablet {
            @include font(28, 52);
        }
        @include laptop {
            @include font(32, 56);
        }
    }
    &.xl {
        @include font-inter-md;
        @include font(26, 54);
        &.rich-text + &.rich-text {
            margin-top: rem(34);
        }
        @include tablet {
            @include font(32, 66);
        }
    }
    &.secondary {
        @include font-inter-md;
    }
    &.bg {
        &-aqua,
        &-white,
        &-pink {
            position: relative;
            display: inline-block;
            padding: rem(2) rem(6);
            line-height: rem(24);
            margin-right: rem(6);
            @include mobileLg {
                padding: rem(2) rem(8);
                margin-right: rem(4);
            }
        }
        &-aqua {
            @include boxShadow(color(pastel, 'aqua'));
        }
        &-white {
            @include boxShadow(color(common, 'white'));
        }
        &-pink {
            @include boxShadow(color(pastel, 'pink'));
        }
    }
}
