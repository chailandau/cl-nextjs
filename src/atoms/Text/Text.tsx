import classNames from 'classnames';
import { FC, ReactNode } from 'react';

import styles from './Text.module.scss';

export const textTags = ['p', 'span'] as const;

export const textSizes = ['lg', 'md', 'sm', 'xs'] as const;
interface TextProps {
    /** Semantic tag text is wrapped in */
    as?: (typeof textTags)[number];
    /** Text content */
    children: string | ReactNode;
    /** Text size */
    size?: (typeof textSizes)[number];
}

const Text: FC<TextProps> = ({ as: TextTag = 'p', size = 'sm', children }) => {
    const classList = classNames(styles['text'], styles[size]);

    return <TextTag className={classList}>{children}</TextTag>;
};

export default Text;
