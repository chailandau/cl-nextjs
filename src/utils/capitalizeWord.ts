/**
 * Capitalizes the first letter of a string.
 *
 * @param {string} str - String to be capitalized.
 * @return {string} Capitalized string.
 */
export const capitalizeWord = (str: string): string => {
    if (!str) {
        return str;
    }

    const firstLetter = str.charAt(0).toUpperCase();
    const restOfWord = str.slice(1);

    return firstLetter + restOfWord;
};
