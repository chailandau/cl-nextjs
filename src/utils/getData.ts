import { graphQLClient } from '@/api/graphqlClient';
import { Query } from '@/api/graphqlTypes';

export const getData = async (QUERY: string) =>
    await graphQLClient.request<Query>(QUERY);

export const getPageData = async (QUERY: string, slug: string) =>
    await graphQLClient.request<Query>(QUERY, { slug });
