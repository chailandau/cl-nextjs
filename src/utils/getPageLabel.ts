import { Nav_MenuItems } from '@/api/graphqlTypes';

export const getPageLabel = (menuItem: Nav_MenuItems) => {
    const customLabel =
        menuItem?.overridePageName &&
        menuItem?.customLabel !== null &&
        menuItem.customLabel;

    return menuItem?.page && (customLabel || menuItem?.page?.title);
};
