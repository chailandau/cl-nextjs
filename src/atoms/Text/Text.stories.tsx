import type { Meta, StoryObj } from '@storybook/react';

import Text, { textSizes, textTags } from './Text';

const meta: Meta<typeof Text> = {
    title: 'Atoms/Text',
    component: Text,
    argTypes: {
        as: {
            options: textTags,
            table: { defaultValue: { summary: 'p' } }
        },
        size: {
            options: textSizes,
            table: { defaultValue: { summary: 'md' } },
            control: { type: 'select' }
        }
    }
};

export default meta;

type Story = StoryObj<typeof Text>;

export const Default: Story = {
    args: {
        as: 'p',
        size: 'md',
        children: 'This is default text (medium).'
    }
};

export const ExtraSmall: Story = {
    args: {
        as: 'p',
        size: 'xs',
        children: 'This is extra small text.'
    }
};
export const Small: Story = {
    args: {
        as: 'p',
        size: 'sm',
        children: 'This is small text.'
    }
};

export const Large: Story = {
    args: {
        as: 'p',
        size: 'lg',
        children: 'This is large text.'
    }
};

export const ExtraLarge: Story = {
    args: {
        as: 'p',
        size: 'xl',
        children: 'This is extra large text.'
    }
};
