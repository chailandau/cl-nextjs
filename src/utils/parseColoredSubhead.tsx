import { JSX } from 'react';

import Text, { TextProps } from '@/atoms/Text';

interface ChildSegment {
    text: string;
    bold?: boolean;
}

export interface ColoredSubheadProps {
    children: ChildSegment[];
}

type ColoredSubhead = JSX.Element[] | null | undefined;

/**
 * Parses a colored subhead and returns an array of Text components with appropriate styling.
 *
 * @param coloredSubhead - An array of objects representing the colored subhead.
 * @returns An array of Text components with appropriate styling, or null if the input is falsy.
 */
export const parseColoredSubhead = (
    coloredSubhead: ColoredSubheadProps[]
): ColoredSubhead | null => {
    if (!coloredSubhead) {
        return null;
    }

    const colors = ['aqua', 'white', 'pink'];

    return coloredSubhead.flatMap((segment) => {
        const boldIndexes: number[] = [];

        return segment?.children?.map((child, index) => {
            const isBold = child?.bold;
            let textBgColor;

            if (isBold) {
                boldIndexes.push(index);
                const position = boldIndexes?.indexOf(index);
                textBgColor = colors[position % colors?.length];
            }

            const textProps = {
                as: 'span' as TextProps['as'],
                size: 'md' as TextProps['size'],
                font: 'secondary' as TextProps['font'],
                backgroundColor: textBgColor
            };

            return (
                <Text key={index} {...textProps}>
                    {child?.text}
                </Text>
            );
        });
    });
};
