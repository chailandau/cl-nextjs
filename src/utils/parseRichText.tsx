import { JSX } from 'react';

import styles from '../atoms/Text/Text.module.scss';

import Text from '@/atoms/Text';
import { textSizes } from '@/atoms/Text/Text';

interface ChildSegment {
    text: string;
}

export interface RichTextProps {
    children: ChildSegment[];
}

type RichText = JSX.Element[] | null | undefined;

/**
 * Parses rich text and returns an array of Text components.
 *
 * @param richText - An array of objects representing rich text.
 * @returns An array of Text components, or null if the input is falsy.
 */
export const parseRichText = (
    richText: RichTextProps[],
    size: (typeof textSizes)[number] = 'xs'
): RichText | null => {
    if (!richText) {
        return null;
    }

    const results: JSX.Element[] = [];

    richText.flatMap(
        (segment) =>
            segment?.children?.forEach((child) => {
                const paragraphs = child.text
                    .split('\n')
                    .filter((paragraph) => paragraph.trim() !== '');

                paragraphs.forEach((par, parIndex) => {
                    results.push(
                        <Text
                            className={styles['rich-text']}
                            size={size}
                            key={parIndex}
                        >
                            {par}
                        </Text>
                    );
                });
            })
    );

    return results;
};
