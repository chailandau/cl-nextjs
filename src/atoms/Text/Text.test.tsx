import { composeStory } from '@storybook/react';

import { textTags } from './Text';
import Meta, { Default } from './Text.stories';

import { testAxeAndSnapshot, testTags } from '@/utils/testHelpers';

const Text = composeStory(Default, Meta);

const textComponent = {
    component: <Text />
};

describe('Text', () => {
    testTags({ ...textComponent, tags: textTags });
    testAxeAndSnapshot(textComponent);
});
