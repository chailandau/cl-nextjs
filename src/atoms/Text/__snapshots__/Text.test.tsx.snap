// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="text md primary"
  >
    This is default text (medium).
  </p>
</DocumentFragment>
`;
